### ISSUE TITLE
Manage sets of dependencies via Gradle or another tool

### ISSUE THREAD

**Issue by cecilia-donnelly** [SolutionGuidance/psm#16]
@jasonaowen [writes](https://groups.google.com/d/msg/psm-dev/nzC3pP1XMFg/MOCUKnkhBAAJ):

> From the `psm-app` directory, you can run `ant deploy` to build an EAR file. It depends on libraries provided by the application server - this is potentially something to change later - and my WIP branch pulls those libraries in. 

Rather than depending on Wildfly to provide libraries, manage dependencies via a few different `build.xml`s or use [Gradle](https://gradle.org/).  For example:

- `wildfly-config/build.xml` points at Wildfly libraries
- `websphere-config/build.xml` downloads stuff from [Maven](https://maven.apache.org/)

The exact approach will be fleshed out as we go.  This issue is for discussion and capturing progress.

**Comment by cecilia-donnelly** [comment]
See also https://github.com/OpenTechStrategies/psm/pull/27/commits/5920b76130689fc6264fd6bbe28ee680a4bc70a2.

**Comment by jasonaowen** [comment]
We should also, as part of this effort, update the JavaScript build process in the same way, as we have several vendored libraries in [cms-web/WebContent/js](https://github.com/OpenTechStrategies/psm/tree/master/psm-app/cms-web/WebContent/js).

**Comment by slifty** [comment]
I think we can close this issue at this point -- @cecilia-donnelly I'll let you do it, or if there are additional actions that need to be taken please specify them.

**Comment by jasonaowen** [comment]
I count two open questions we haven't addressed:

1. What do we do about vendored JavaScript libraries? We can address that in a separate issue if we feel that this one has gotten too big / long, but if so we should open that new issue before closing this one.
2. What do we do about dependencies provided by WildFly? We had talked about bulk-importing WildFly's libraries directory as a first step, and refining that to specific [`compileOnly` dependencies](https://blog.gradle.org/introducing-compile-only-dependencies) later. Is that still the plan?

**Comment by jasonaowen** [comment]
I've been thinking a bit about how we should handle the JavaScript libraries we currently have vendored. I see two options:

1. Use [npm](https://www.npmjs.com/). This is the standard JavaScript packaging system, and its framework would allow us to do more sophisticated things later, including asset compilation to speed up page load times. However, it has the drawback of being an entirely separate system from Java & gradle. We will probably want gradle to know how to invoke npm, so that we have only a single build step, but it may be challenging to integrate - we'll want to use npm to stage our JavaScript dependencies, and then pull them into the war file with gradle.

2. Use [WebJars](http://www.webjars.org/). WebJars package JavaScript libraries into jar files that we could depend on in gradle and whose contents are put in a well-known location in the war file. We'd most likely need to update Spring to know how to route requests for such assets. This has the advantage of being fairly straightforward to implement, but has the drawback of being a nonnative JavaScript system, which could make future contributors less effective at working on the front end.

Both approaches would make it easier for us to upgrade our JavaScript libraries, and give us more visibility into the versions we're using and their licenses.

Do any of you have thoughts on this? Have any of you used either approach in a Java + JavaScript project? How did it go?

**Comment by cecilia-donnelly** [comment]
I'll be talking about this tomorrow with @slifty, @jasonaowen, and @PaulMorris.  Jason set up [a versioneye account](https://www.versioneye.com/user/projects/597b66db6725bd0040fda957) to track the status of our dependencies.

I went through the ones that versioneye flags as being out of date and made notes:

1. jBPM is its own thing: My understanding is that we're going to update the separate business rules engine separately, so I'm not worrying about this right now. (See #385)
1. Groovy: We're on 2.4, which is the latest stable version, but could go up a couple minor versions (I don't see any breaking changes) http://groovy-lang.org/changelogs/changelog-2.4.12.html
1. Serenity-core: Our current version is a few months old, probably easy to upgrade but not necessary
1. jaxb: https://javaee.github.io/jaxb-v2/doc/user-guide/ch02.html
    - security fixes are named in all the minor version changes, so this is probably worth updating.  They don't call out any breaking changes, but that's no guarantee of course.
1. commons-codec
    - 1.6 is from 2011
    - This is a good candidate for upgrade -- no breaking changes called out in https://commons.apache.org/proper/commons-codec/changes-report.html
    - We'd need Java 1.6 at a minimum, but I think we're there
1. commons-io 
    - 2.0.1 is from 2010
    - 2.2 is the oldest version available on their [download site]https://commons.apache.org/proper/commons-io/index.html), so I suspect we should try to at least get there
1. commons-lang 
    - They've had a major version upgrade.  [Notes are here](http://commons.apache.org/proper/commons-lang/article3_0.html)
    - Upgrading this a major version may be too big of a lift -- thoughts welcome
1. hapi-fhir-client-okhttp 
    - They have moved from 2.5 to 3.0, but it looks like we're good to stay on 2.5 for now (it's still new).  Moving to 3.0 looks like it would entail quite a bit of [rewriting](http://hapifhir.io/changes-report.html)
1. itext 
    -  See #392
    - Or switch to OpenPDF
1. jasypt
    - Do we need this in addition to bouncycastle?
    - It looks like this is used in Util.java and cms.properties to encrypt prod properties
    - Upgrading this along with Spring seems like a good idea: http://www.jasypt.org/whatsnew19.html
1. log4j 
    - The latest 1.2.x maintenance release was in 2012
    - They've moved up a major version to 2 -- there is a migration guide [here](https://logging.apache.org/log4j/2.x/manual/migration.html)
    - Opinions about how important it is to upgrade?
1. jxl 
    - I don't think we need this.
1. janino 
    - I don't think we need this.
    - http://janino-compiler.github.io/janino/changelog.html
1. slf4j-api and slf4j-log4j12 
    - They still offer 1.6.4 on their download page:
    https://www.slf4j.org/dist/
    - It came out in 2011.  This would need to be upgraded in parallel with log4j, if at all.

More to come tomorrow.

**Comment by slifty** [comment]
I'll take on the commons libraries (5,6,7 from the above list) and jaxb (4) this weekend.
We discussed the logging dependencies as well and decided that #91 should be taken on first, and we need to decide whether log4j or java util is the right way forward there..

**Comment by jasonaowen** [comment]
Okay, I've done a bit of reading and research, and here's what I've found.

First, we're including `log4j`, but we only use it directly in three places, in addition to using it through the TopCoder helper library. It's also included via `jxl`, which we don't seem to use anywhere; `jxl` is on the list of things that were vendored and put into gradle with the intention of reexamining them (in this issue!) - I agree with @cecilia-donnelly's assessment that we can remove `jxl` safely.

Second, one of the options we have is to use [SLF4J](https://www.slf4j.org/), a logging facade. My understanding is that this is very useful when writing a library, as you don't want to introduce a logging engine to your users; instead, by using SLF4j you allow your users to configure their choice of logging engine at deploy time. As an application rather than a library, this does not apply to us, and so I suggest we not use SLF4J. (See also [these](https://softwareengineering.stackexchange.com/a/67424/202259) [SE](https://softwareengineering.stackexchange.com/a/108686/202259) [questions](https://stackoverflow.com/a/11360517/25637).)

So, the question in my mind is whether to use `java.util.logging` (JUL), [Log4j 2](https://logging.apache.org/log4j/2.x/), or [Logback](https://logback.qos.ch/). This seems to be a subjective question, with the only non-subjective aspect being the performance of the logging engine. This may be a failure of my imagination, but I do not foresee, in this application, the logging engine becoming a bottleneck.

The primary benefit of JUL is that it is built-in. That means one fewer dependency to worry about and upgrade as new versions are released, which seems valuable to me in this project. Secondarily, [there are adapters from JUL to SLF4J](https://stackoverflow.com/a/20407321/25637), and from there you can configure other engines - effectively, then, JUL is both an engine and a facade.

In addition to performance, there are a number of features that each of Log4j and Logback have, and the most potentially relevant to us, I think, is advanced configuration options, including detailed filtering and highly specific log setups.

That said, looking through the feature lists on [the Log4j 2 website](https://logging.apache.org/log4j/2.x/) and [the Logback website](https://logback.qos.ch/), nothing seemed particularly compelling to me, nor enough to justify the additional dependency. My preference would be to use the built-in `java.util.logging`, but I am open to persuasion if there is something someone thinks I missed or undervalued.

Additional reading:
- [What are the latest options in Java logging frameworks?](https://stackoverflow.com/a/2533856/25637)
- [Is Standard Java Logging Dead? Log4j vs Log4j2 vs Logback vs java.util.logging](http://blog.takipi.com/is-standard-java-logging-dead-log4j-vs-log4j2-vs-logback-vs-java-util-logging/)

**Comment by cecilia-donnelly** [comment]
Still to do:
- Upgrade Groovy, itext, jasypt
- Remove jxl, janino

**Comment by cecilia-donnelly** [comment]
@jasonaowen, this is only waiting for review/merge of #630, right?

**Comment by jasonaowen** [comment]
@cecilia-donnelly right!
